Add better error message when trying to get edge that is not present

finsberg opened the pull request:
<!--
Please use pre-commit to lint your code.
For more details check out step 1 and 4 of
https://networkx.org/documentation/latest/developer/contribute.html
-->

rossbar reviewed (commented):
I think a better exception message would be an improvement, though changing the exception type would be a backwards-incompatible change. One possible compromise would be to re-raise the `KeyError` with the new exception message (i.e. replace the `nx.NetworkXError` in this PR with `KeyError`).

This would give the better exception message while preserving back compat - WDYT?

finsberg commented:
> I think a better exception message would be an improvement, though changing the exception type would be a backwards-incompatible change. One possible compromise would be to re-raise the `KeyError` with the new exception message (i.e. replace the `nx.NetworkXError` in this PR with `KeyError`).
> 
> This would give the better exception message while preserving back compat - WDYT?

Or, if it is important that all exceptions that are emitted from NetworkX is of type `nx.NetworkXError`, we could create a new exception that is a subclass of `KeyError` and `nx.NetworkXError`, i.e
```python
class NetworkXKeyError(KeyError, nx.NetworkXError)
    pass
```

rossbar commented:
> if it is important that all exceptions that are emitted from NetworkX is of type 

IMO this isn't important, but perhaps others have more informed opinions. I'd prefer to stick with builtin exceptions where possible, and *I think* KeyError works nicely for this case (NX graphs are "dicts all the way down" after all). I can be persuaded otherwise though; the only thing I'm -1 on is a backward-incompatible change in the exception type!

finsberg commented:
Sounds good. I changed it to re-raise `KeyError`. Also casting the exception to a string didn't work (see https://github.com/python/cpython/issues/116408#issuecomment-1980430949) so that is why I changed it to use `e.value.args` instead.

dschult commented on networkx/classes/tests/test_reportviews.py:
```diff
@@ -596,8 +596,18 @@ def test_getitem(self):
+        )
+
+        # Invalid edge
+        with pytest.raises(KeyError) as e:
+            G.edges[0, 9]
+
+        assert e.value.args[0] == "The edge (0, 9) is not in the graph."
```
There is an argument to pytest.raises that allows you to test matching the error message:
```suggestion
        with pytest.raises(KeyError, match="The edge (0, 9) is not in the graph."):
            G.edges[0, 9]
```
The same can be done to the first test.  And the match argument doesn't need to exactly match. It can be a substring, and it also accepts/uses regular expressions. See the pytest.raises docs if you are interested.

finsberg replied:
Very nice! Didn't know about that one.

rossbar reviewed (approved):
I took the liberty of pushing a few changes (mostly cosmetic) and tried to simplify the `match=` logic a bit by making more liberal use of wildcards.

Otherwise this LGTM, thanks @finsberg !

dschult reviewed (approved):
Looks good... Thanks!
